hummingbot/hummingbot
# refactor / binance perpetual connector

**Before submitting this PR, please make sure**:


**A description of the changes proposed in the pull request**:

- REST API & sockets -> web_assistant
- Single websocket connection in order book data source with queues for diffs and trades
- Throttler
- Client Orders Tracker
- Token-pair mapping from exchange
- Docstrings for interface methods

**Tests performed by the developer**:

- Ran the unit tests
- Ran perpetual market making with the connector on testnet

**Tests performed by QA**:

- Checked the API key connection and balance update
- Created a perpetual market making strategy and test import option
- Confirmed that trading pairs are getting fetched
- Checked that order price is synchronization
- Checked the trade fill events
- Run test bots for long-duration using different markets to capture the funding details: Confirmed that funding details are displayed without any error
- Run a spot-perpetual arbitrage strategy to check binance perpetual connector's behavior when processing a market order
- Several issues have been reported and had been addressed by the dev team.
- update: update long-term test bot whenever a new commit is pushed and resume long-term test. For couple of days, we did not hit any new error.
